### ISSUE TITLE
ShaderTerrainMesh performance

### ISSUE THREAD

**Issue by thetestgame** [panda3d/panda3d#951]
I was unable to determine if this is a true issue or just part of my ignorance to how internally panda works. I was doing some experimenting in my free time to solve a personal curiosity when I noticed that the ShaderTerrainMesh is sending what looks like its height field texture to the GPU every frame. although its not much at first however when you start adding more and more ShaderTerrainMeshes it rapidly gets out of hand as shown by my PStats GIF. Is this normal behavior? 

https://i.gyazo.com/30042ae04a2bdab19bd3b01928b8f6ac.gif

**Comment by thetestgame** [comment]
I looked back in an old conversation with @rdb and noticed I had found and discussed this issue with him back in the start of 2019. It does very much seem to be a bug.

**Comment by thetestgame** [comment]
I've created a simple example that for me showcases the issue in PStats fairly well. 

![image](https://i.gyazo.com/d380ef801b7b85b0a2cc6c4838418eb9.gif)
[Terrain.zip](https://github.com/panda3d/panda3d/files/4747423/Terrain.zip)

**Comment by el-dee** [comment]
Each frame the ShaderTerrainMesh generate a data texture that contains the configuration of each chunk to display and this for a fixed number of views, that explain the data transfer you observe. 

You can reduce the size of the data texture by using the PRC variables : 

stm-max-views : By default STM assumes you have up to 8 cameras, if you have less you can reduce this number (If you have only one camera you can set it to 1 and divide the amount of data by 8)

stm-max-chunk-count : Number of chunk per terrain, the default value is 2048, you can reduce it but it will lessen the quality.

**Comment by rdb** [comment]
Indeed, I had intended to comment what @el-dee just pointed out.  This isn't the heightfield being transferreed, but the data texture.  The defaults are rather rich and can be reduced a lot.

@thetestgame: on Discord you pointed out that you are still seeing transfers for terrains that are entirely out of view.  This could potentially be improved, but we'd appreciate a simple test case for that.

**Comment by Moguri** [comment]
Is there something actionable here (e.g., better defaults), or can this be closed?
